src/main/java/io/cryostat/net/web/http/api/beta/MatchExpressionsPostHandler.java

```diff
@@ -0,0 +1,183 @@
+import io.cryostat.net.web.http.api.v2.IntermediateResponse;
+import io.cryostat.net.web.http.api.v2.RequestParameters;
+import io.cryostat.platform.ServiceRef;
+import io.cryostat.rules.MatchExpression;
+import io.cryostat.rules.MatchExpressionManager;
+import io.cryostat.rules.MatchExpressionManager.MatchedMatchExpression;
+import io.cryostat.rules.MatchExpressionValidationException;
+
+import com.google.gson.Gson;
+import io.vertx.core.http.HttpHeaders;
+import io.vertx.core.http.HttpMethod;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.hibernate.exception.ConstraintViolationException;
+
+public class MatchExpressionsPostHandler extends AbstractV2RequestHandler<MatchedMatchExpression> {
+
+    static final String PATH = "matchExpressions";
+
+    private final MatchExpressionManager expressionManager;
+    private final NotificationFactory notificationFactory;
+
+    @Inject
+    MatchExpressionsPostHandler(
+            AuthManager auth,
+            CredentialsManager credentialsManager,
+            MatchExpressionManager expressionManager,
+            NotificationFactory notificationFactory,
+            Gson gson) {
+        super(auth, credentialsManager, gson);
+        this.expressionManager = expressionManager;
+        this.notificationFactory = notificationFactory;
+    }
+
+    @Override
+    public boolean requiresAuthentication() {
+        return true;
+    }
+
+    @Override
+    public ApiVersion apiVersion() {
+        return ApiVersion.BETA;
+    }
+
+    @Override
+    public HttpMethod httpMethod() {
+        return HttpMethod.POST;
+    }
+
+    @Override
+    public Set<ResourceAction> resourceActions() {
+        return EnumSet.of(ResourceAction.CREATE_MATCH_EXPRESSION);
+    }
+
+    @Override
+    public String path() {
+        return basePath() + PATH;
+    }
+
+    @Override
+    public List<HttpMimeType> produces() {
+        return List.of(HttpMimeType.JSON);
+    }
+
+    @Override
+    public List<HttpMimeType> consumes() {
+        return List.of(HttpMimeType.MULTIPART_FORM, HttpMimeType.URLENCODED_FORM);
+    }
+
+    @Override
+    public boolean isAsync() {
+        return false;
+    }
+
+    @Override
+    public boolean isOrdered() {
+        return true;
+    }
+
+    @Override
+    public IntermediateResponse<MatchedMatchExpression> handle(RequestParameters params)
+            throws ApiException {
+        String matchExpression = params.getFormAttributes().get("matchExpression");
+        String targets = params.getFormAttributes().get("targets");
+        if (StringUtils.isBlank(matchExpression)) {
+            throw new ApiException(400, "'matchExpression' is required.");
+        }
+        try {
+            if (StringUtils.isNotBlank(targets)) {
+                Set<ServiceRef> matched;
+                List<String> parsedTargets = this.expressionManager.parseTargets(targets);
```

Comment:
I'm not sure about this - what are the reasons for having the client send a list of `connectUrl`s instead of sending a list of whole `ServiceRef` (Target) objects?

Reply:
Hmm... I did it this way because it was easier for a client to supply these since connectUrls should be unique, but it makes sense that if the web client is sending queries with known structured ServiceRef objects, that wouldn't make that much difference.

Reply:
> Note: It's also a little broken, anyways since connectUrls may not match a Target, jvmIds would make more sense.

This could lead to a circular dependency. The match expression being added/tested could be for a stored credential, in which case it's likely that Cryostat has never been able to connect to the target before to determine its jvmId.